Docs for Asset Manager and Data Config (#77)
wintercms/docs: snowboard-extras.md

        this.config = this.snowboard.config(this, element);
    }

    // ...

}
```

<a name="data-config-methods"></a>
### Methods

The config instance that is returned by `this.snowboard.config()` provides the following methods for interacting with the configuration of a plugin instance.

#### `get()`

Gets the entire configuration as an object, with the configuration name as the object keys and the values as the object values. This object will be made up of the defaults, merged together with the values of the data attributes, which take precedence.

#### `get(configName: string)`

Gets the configuration value for the given configuration name. This will be retrieved from the data attribute of the element providing the configuration, or from the the defaults if not specified on the element.

This will return `undefined` if there is no configuration value in the element's data attributes or the defaults.

#### `set(configName: string, configValue: any, persist: boolean)`

Sets a configuration value for the given configuration name at run-time. This will act as an override - it will replace the data configuration and any default for the configuration option.

> By default, this override will not persist - if the data configuration is refreshed, it will be replaced by the data attribute value or default once more. If the third parameter `persist` is set to `true`, this value will be persisted, and will be kept even if refreshed.

Reviewer: Does the persist flag mean that the value is set in the attribute on the element itself rather than just the data layer?

Author: Yep, correct.

        ];
    }
}
```

You might then have a default (or overwritten) partial which contains the HTML that will be used by your Snowboard plugin.

```twig
<div
    class="gallery"
    data-gallery
    data-num-images="{{ __SELF__.settings().numImages }}"
    data-show-caption="{{ __SELF__.settings().showCaption }}"
>
    <img src="picture-1.png" title="This is a cool picture">
    <img src="picture-2.png" title="This is another cool picture">
    <img src="picture-3.png" title="Look at this">
    <img src="picture-4.png" title="Wow, so cool!">
    <img src="picture-5.png" title="Nice!">
</div>
```

With this in place, your Snowboard plugin can now initialise a gallery with configuration direct from the Backend.

```js
class Gallery extends Snowboard.PluginBase {

    constructor(snowboard, element) {
        super(snowboard);
        this.element = element;

        // Initialise the configuration, and make it available in the plugin as "this.config"
        this.config = this.snowboard.config(this, element);

        this.createGallery();
    }

    defaults() {
        return {
            numImages: 5,
            showCaption: false,
        };
    }

    createGallery() {

>         const numImages = this.config.get('numImages'); // Will return 3, as per the default component config

Reviewer: ```suggestion
        const numImages = this.config.get('numImages'); // Will return 3 since the attribute has been set by the default value defined in the PHP component class
```

>         const showCaption = this.config.get('showCaption'); // Will be true, as per the default component config

Reviewer: ```suggestion
        const showCaption = this.config.get('showCaption'); // Will be true, since the attribute has been set by the default value defined in the PHP component class
```

    }
}
```

Following this structure, you have full-stack control over the experience of your component, providing an easy mechanism for controlling the frontend widget from the backend.

<a name="data-config-usage"></a>
### Usage

Initialising a data configuration requires two parameters, the Snowboard plugin that you wish to make the config available to, and a HTML element to extract the data configuration from.

```js
this.config = this.snowboard.config(

>     this, // Use the current instance to add the config to
>     element, // This should be the HTML element that you wish to get the config from

Reviewer: ```suggestion
    this, // Add the config to the current instance
    element, // HTML element to get the config values from
```

);
```

By populating it to a plugin variable, you can use it throughout the plugin.

> **NOTE:** The configuration keys (ie. the name from the data attribute name) follow the [name conversion](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/dataset#name_conversion) methodology of the HTML Element dataset. This means, in general, that names will be converted to "camelCase" without the `data-` prefix. In the above example, `data-num-images` is converted to `numImages` within the configuration on the JavaScript side.

When determining the available configuration options, the data configuration will look for a `defaults` method in the plugin instance. This method must return an object that has the accepted configuration options as the object keys, and the default values of these options as the object values.

```js
class Gallery extends Snowboard.PluginBase {

    // ...

    defaults() {
        return {
            numImages: 5,
            showCaption: false,
        };
    }

    // ...

}

> ```

Reviewer: @bennothommo What's the point of specifying default values in the JS if they're also specified in the PHP?

Author: @LukeTowers there's two reasons - one, the `defaults` method sets the available fields that can be populated from the data attributes, to prevent people from adding arbitrary config data unless they specifically allow it. The second is that the widget or function might work independently from a component but still use the data configuration system.

In the above example, this would allow the `data-num-images` and `data-show-caption` attributes in the given element to populate the configuration, however, if another data attribute (such as `data-allow-zoom`) were added, this would *not* be available as a configuration option.

If you would like to accept any data attribute as a configuration value, you may instead add a property to the plugin instance called `acceptAllDataConfigs` with the value as `true`.

```js
class Gallery extends Snowboard.PluginBase {

    constructor(snowboard, this, element) {
        super(snowboard);

        this.element = element;
        this.acceptAllDataConfigs = true;
        this.config = this.snowboard.config(this, element);
    }

    // ...

}
```

<a name="data-config-methods"></a>
### Methods

> The config instance that is returned by `this.snowboard.config()` provides the following methods for interacting with the configuration of a plugin instance.

Reviewer: ```suggestion
The configuration instance that is returned by `this.snowboard.config(bindTo, elementFrom)` provides the following methods:
```

#### `get()`

Gets the entire configuration as an object, with the configuration name as the object keys and the values as the object values. This object will be made up of the defaults, merged together with the values of the data attributes, which take precedence.

#### `get(configName: string)`

Gets the configuration value for the given configuration name. This will be retrieved from the data attribute of the element providing the configuration, or from the the defaults if not specified on the element.

This will return `undefined` if there is no configuration value in the element's data attributes or the defaults.

#### `set(configName: string, configValue: any, persist: boolean)`

Sets a configuration value for the given configuration name at run-time. This will act as an override - it will replace the data configuration and any default for the configuration option.

By default, this override will not persist - if the data configuration is refreshed, it will be replaced by the data attribute value or default once more. If the third parameter `persist` is set to `true`, this value will be persisted, and will be kept even if refreshed.

#### `refresh()`

> Refreshes the entire configuration. This will repopulate the configuration with the data attribute configuration or default values once more. This can be useful if you allow the data attributes of the element to be modified externally.

Reviewer: Can you add small usage examples to each of the methods here?

    defaults() {
        return {
            numImages: 5,
            showCaption: false,
        };
    }

    createGallery() {
        const numImages = this.config.get('numImages'); // Will return 3, as per the default component config
        const showCaption = this.config.get('showCaption'); // Will be true, as per the default component config
    }
}
```

> Following this structure, you have full-stack control over the experience of your component, providing an easy mechanism for controlling the front-end widget from the backend.

Reviewer: ```suggestion
Following this structure, you have full-stack control over the experience of your component, providing an easy mechanism for controlling the frontend widget from the backend.
```
